tpm2tools/keys.go

```diff
@@ -133,3 +135,163 @@ func (k *Key) PublicKey() crypto.PublicKey {
+// produced by the PCR state in pcrs. Similar to seal and unseal, this acts on
+// the SHA256 PCRs and uses the ownerhierarchy.
+// The Key k is used as the parent key.
+func (k *Key) Reseal(rw io.ReadWriter, pcrs map[int][]byte, priv, pub []byte) ([]byte, []byte, error) {
+	pcrNums := make([]int, 0, len(pcrs))
+	for key := range pcrs {
+		pcrNums = append(pcrNums, key)
+	}
+
+	sensitive, err := k.Unseal(rw, pcrNums, priv, pub)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unseal: %v", err)
+	}
+
+	auth, err := computePCRSessionAuth(rw, pcrs)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to compute pcr session auth: %v", err)
+	}
+
+	return tpm2.Seal(rw, k.Handle(), "", "", auth, sensitive)
+}
+
+type tpmsPCRSelection struct {
+	Hash tpm2.Algorithm
+	Size byte
+	PCRs tpmutil.RawBytes
+}
+
+type sessionSummary struct {
+	OldDigest      tpmutil.RawBytes
+	CmdIDPolicyPCR uint32
+	NumPcrSels     uint32
+	Sel            tpmsPCRSelection
+	PcrDigest      tpmutil.RawBytes
+}
+
+func computePCRSessionAuth(rw io.ReadWriter, pcrs map[int][]byte) ([]byte, error) {
+	pcrBits := make([]byte, 3)
+	for pcr := range pcrs {
+		byteNum := pcr / 8
+		bytePos := byte(1 << byte(pcr%8))
+		pcrBits[byteNum] |= bytePos
+	}
+	pcrDigest := digestPCRList(pcrs)
+
+	summary := sessionSummary{
+		OldDigest:      make([]byte, sha256.Size),
+		CmdIDPolicyPCR: uint32(0x0000017f),
+		NumPcrSels:     uint32(1),
+		Sel: tpmsPCRSelection{
+			Hash: tpm2.AlgSHA256,
+			Size: uint8(3),
+			PCRs: pcrBits,
+		},
+		PcrDigest: pcrDigest,
+	}
+	b, err := tpmutil.Pack(summary)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack for hashing: %v ", err)
+	}
+
+	hash := crypto.SHA256.New()
+	hash.Write(b)
+	return hash.Sum(nil), nil
+}
+
+func digestPCRList(pcrs map[int][]byte) []byte {
+	hash := crypto.SHA256.New()
+	for i := 0; i < 24; i++ {
+		_, exists := pcrs[i]
+		if !exists {
+			continue
+		}
+		binary.Write(hash, binary.BigEndian, pcrs[i])
+	}
+	return hash.Sum(nil)
+}
+
+func getPCRSessionAuth(rw io.ReadWriter, pcrs []int) ([]byte, error) {
+	handle, err := createPCRSession(rw, pcrs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get digest: %v", err)
+	}
+	defer tpm2.FlushContext(rw, handle)
+
+	digest, err := tpm2.PolicyGetDigest(rw, handle)
+	if err != nil {
+		return nil, fmt.Errorf("could not get digest from session: %v", err)
+	}
+
+	return digest, nil
+}
+
+func createPCRSession(rw io.ReadWriter, pcrs []int) (tpmutil.Handle, error) {
+	nonceIn := make([]byte, 16)
+	/* This session assumes the bus is trusted.  */
+	handle, _, err := tpm2.StartAuthSession(
+		rw,
+		tpm2.HandleNull,
+		tpm2.HandleNull,
+		nonceIn,
+		nil,
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		tpm2.AlgSHA256)
+
+	if err != nil {
+		return tpm2.HandleNull, fmt.Errorf("failed to start auth session: %v", err)
+	}
+
+	sel := tpm2.PCRSelection{
+		Hash: tpm2.AlgSHA256,
+		PCRs: pcrs,
+	}
+	err = tpm2.PolicyPCR(rw, handle, nil, sel)
```

Comment:
Here and elsewhere, combine like:

if err := tpm2.PolicyPCR(rw, handle, nil, sel); err != nil {
  return tpm2.HandleNull, fmt.Errorf("auth step PolicyPCR failed: %v", err)
}

Reply:
Done.